File: Graph/ts/prereqsPossible.ts
export function prereqsPossible(numCourses: number, prereqs: Array<[number, number]>): boolean {
  const graph: Record<number, number[]> = {};
  for (let i = 0; i < numCourses; i += 1) {
    graph[i] = [];
  }
  for (const [a, b] of prereqs) {
    graph[a].push(b);
  }

  const visiting = new Set<number>();
  const visited = new Set<number>();

  for (let node = 0; node < numCourses; node += 1) {
    if (hasCycle(graph, node, visiting, visited)) {
      return false;
    }
  }

  return true;
}

function hasCycle(graph: Record<number, number[]>, node: number, visiting: Set<number>, visited: Set<number>): boolean {
  if (visited.has(node)) return false;
  if (visiting.has(node)) return true;

  visiting.add(node);
  for (const neighbor of graph[node]) {
    if (hasCycle(graph, neighbor, visiting, visited)) return true;
  }
  visiting.delete(node);
  visited.add(node);
  return false;
}

File: Graph/go/prereqs_possible.go
package graph

// PrereqsPossible returns whether all courses can be completed (i.e., no cycles).
func PrereqsPossible(numCourses int, prereqs [][2]int) bool {
	graph := make(map[int][]int, numCourses)
	for i := 0; i < numCourses; i++ {
		graph[i] = []int{}
	}
	for _, edge := range prereqs {
		a, b := edge[0], edge[1]
		graph[a] = append(graph[a], b)
	}

	visiting := map[int]struct{}{}
	visited := map[int]struct{}{}

	for node := 0; node < numCourses; node++ {
		if prereqCycle(graph, node, visiting, visited) {
			return false
		}
	}
	return true
}

func prereqCycle(graph map[int][]int, node int, visiting, visited map[int]struct{}) bool {
	if _, done := visited[node]; done {
		return false
	}
	if _, active := visiting[node]; active {
		return true
	}

	visiting[node] = struct{}{}
	for _, neighbor := range graph[node] {
		if prereqCycle(graph, neighbor, visiting, visited) {
			return true
		}
	}
	delete(visiting, node)
	visited[node] = struct{}{}
	return false
}
